apps_download/models/product_product.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2017-Today: Odoo Community Association (OCA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'product_product_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        if not self._check_m2m_recursion('dependent_product_ids'):
+            raise ValidationError(
+                _('Error: You cannot create recursive dependency.')
+            )
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
```

Comment:
You could avoid this addtional `if` since that `res |= []` concatenate nothing

Reply:
added |= but IMO if condition will be better so that at least the function will not be called which is not needed in case if there are no dependent products.